### PR TITLE
Fix  COMET_EXPERIMENT_KEY environment variable usage in comet logger

### LINKED ISSUE #4229: Comet logger overrides COMET_EXPERIMENT_KEY env variable
After https://github.com/PyTorchLightning/pytorch-lightning/pull/2553  there is a changed logger behavior. It starts using `COMET_EXPERIMENT_KEY`. But it doesn't respect it if it is set already.
So the bug is in the following.
I already set this variable 
Then logger overwrites my value here https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/loggers/comet.py#L189
Then it deletes this variable at all here https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/loggers/comet.py#L215
This way it ignores my variable and deletes it at all later
Moreover in version function it also ignores my set variable
I will create a pull request to fix it

### PR DESCRIPTION
**Pull request by Vozf**
## What does this PR do?

Fixes #4229 (issue)

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/loggers/comet.py [hunk 1]
```diff
@@ -185,7 +185,8 @@ def experiment(self):
         if self._experiment is not None:
             return self._experiment
 
-        if self._future_experiment_key is not None:
+        manual_set_experiment_key_environment_variable = self._future_experiment_key is not None
+        if manual_set_experiment_key_environment_variable:
```
Comment (Lothiraldan) [inline pytorch_lightning/loggers/comet.py:189]: This looks good, we can also move the cleaning of `self._future_experiment_key` in the finally block, this way we don't need an additional variable.

**Review by Lothiraldan — COMMENTED** [review]
This looks good, we should probably add a test for it.

File: pytorch_lightning/loggers/comet.py (hunk 1 above)
Comment (Vozf) [reply pytorch_lightning/loggers/comet.py:189]: Do you mean `self._future_experiment_key =  os.environ.pop("COMET_EXPERIMENT_KEY")` in finally or something else? I think I don't really follow your idea

Comment (Lothiraldan) [reply pytorch_lightning/loggers/comet.py:189]: No I mean remote the line `self._future_experiment_key = None` from here and replace the finally block by:

```
            if self._future_experiment_key is not None:
                os.environ.pop("COMET_EXPERIMENT_KEY")
                self._future_experiment_key = None
```

**Review by Borda — DISMISSED** [review]
could we pls add a test to show failing master and this is the fixes... 🐰

**Comment by tchaton** [comment]
Hey @Vozf, Could you add a tests by any chance ?

**Comment by Vozf** [comment]
Not really. Made this pr based on my personal inherited cometLogger class and I don't even have a setup for PytorchLightning project. I am also not familiar with your testing setup

**Comment by Lothiraldan** [comment]
I've worked on a test for this PR, it's a bit more complex than needed because the experiment key is not passed as a parameter to the Experiment class, I don't know what is the best way to send it to this PR. Here is the patch if someone with the right permission can push it to this PR:

```diff
diff --git a/tests/loggers/test_comet.py b/tests/loggers/test_comet.py
index dc66f184..3d13c6df 100644
--- a/tests/loggers/test_comet.py
+++ b/tests/loggers/test_comet.py
@@ -12,7 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 import os
-from unittest.mock import patch
+from unittest.mock import patch, DEFAULT
 
 import pytest
 
@@ -25,6 +25,7 @@ from tests.base import EvalModelTemplate
 def _patch_comet_atexit(monkeypatch):
     """ Prevent comet logger from trying to print at exit, since pytest's stdout/stderr redirection breaks it. """
     import atexit
+
     monkeypatch.setattr(atexit, "register", lambda _: None)
 
 
@@ -99,6 +100,37 @@ def test_comet_logger_experiment_name(comet):
         comet_experiment().set_name.assert_called_once_with(experiment_name)
 
 
+@patch('pytorch_lightning.loggers.comet.comet_ml')
+def test_comet_logger_manual_experiment_key(comet):
+    """Test that Comet Logger respects manually set COMET_EXPERIMENT_KEY."""
+
+    api_key = "key"
+    experiment_key = "96346da91469407a85641afe5766b554"
+
+    instantation_environ = {}
+
+    def save_os_environ(*args, **kwargs):
+        nonlocal instantation_environ
+        instantation_environ = os.environ.copy()
+
+        return DEFAULT
+
+    # Test api_key given
+    with patch.dict(os.environ, {"COMET_EXPERIMENT_KEY": experiment_key}):
+        with patch('pytorch_lightning.loggers.comet.CometExperiment', side_effect=save_os_environ) as comet_experiment:
+            logger = CometLogger(api_key=api_key)
+
+            assert logger.version == experiment_key
+
+            assert logger._experiment is None
+
+            _ = logger.experiment
+
+            comet_experiment.assert_called_once_with(api_key=api_key, project_name=None)
+
+    assert instantation_environ["COMET_EXPERIMENT_KEY"] == experiment_key
+
+
 @patch('pytorch_lightning.loggers.comet.CometOfflineExperiment')
 @patch('pytorch_lightning.loggers.comet.comet_ml')
 def test_comet_logger_dirs_creation(comet, comet_experiment, tmpdir, monkeypatch):
```

File: pytorch_lightning/loggers/comet.py [hunk 2]
```diff
@@ -278,6 +279,9 @@ def version(self) -> str:
         if self._experiment_key is not None:
             return self._experiment_key
 
+        if "COMET_EXPERIMENT_KEY" in os.environ:
+            return os.environ["COMET_EXPERIMENT_KEY"]
+
```
Comment (awaelchli) [inline pytorch_lightning/loggers/comet.py:284]: sorry for the question, I am not very familiar with this logger. 
if the env variable is supposed to determine the version, can it not be assigned to self._experiment_key, and then we use the above if statement to return the right version? What's the difference between self._experiment_key and the env variable?

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/loggers/comet.py (hunk 2 above)
Comment (Vozf) [reply pytorch_lightning/loggers/comet.py:284]: logic is based on the value of `self._experiment_key`  for example here https://github.com/PyTorchLightning/pytorch-lightning/blob/357de8e2bc610fcf32f96e5fed97b832493cbde2/pytorch_lightning/loggers/comet.py#L193
tldr if it is set it means that the experiment is already created and you should log in it. If not then create new experiment. Here we want a new experiment but with predefined experiment key, which is set with env variable

Comment (Lothiraldan) [reply pytorch_lightning/loggers/comet.py:284]: Indeed the name is a bit confusing, the env variable `COMET_EXPERIMENT_KEY` is used to manually set the experiment key for new experiment. The `self._experiment_key` attribute is used to "continue" a previous experiment, it's passed as the `previous_experiment` parameter of `CometExistingExperiment`.

If the name don't need to be standard across loggers, `self._experiment_key` could be renamed `self._previous_experiment_key` in order to reduce confusion.

Comment (Vozf) [reply pytorch_lightning/loggers/comet.py:284]: Would be strange renaming internal variable and not renaming the api here. 
https://github.com/PyTorchLightning/pytorch-lightning/blob/357de8e2bc610fcf32f96e5fed97b832493cbde2/pytorch_lightning/loggers/comet.py#L125
If we are renaming to avoid confusion init parameter name should also be renamed to `previous_experiment_key`

Comment (awaelchli) [reply pytorch_lightning/loggers/comet.py:284]: I guess all I was asking is, can we do

```python
self._experiment_key = os.environ.get("COMET_EXPERIMENT_KEY", experiment_key)
```
in init? (or maybe also warn/error to user if both are set)
Because as you explain, there are two ways to continue the experiment, either define the env var or pass in the key to the logger.

Comment (Vozf) [reply pytorch_lightning/loggers/comet.py:284]: As I've said that would affect the logic of the logger and ExistingExperiment will be used in this case which is not expected behaviour. To fix that you need to rethink logic based on `self._experiment_key`
Anyway I don't believe this is in scope of this pr

**Review by Hansel562 — APPROVED** [review]

**Review by tchaton — APPROVED** [review]

**Review by SkafteNicki — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Review by SeanNaren — APPROVED** [review]
